sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementIDisposableCorrectly.cs

```diff
@@ -48,203 +48,239 @@ public class ImplementIDisposableCorrectly : SonarDiagnosticAnalyzer
+            private void VerifyDisposeOverride(MethodDeclarationSyntax disposeMethod)
             {
-                yield break;
-            }
+                if (disposeMethod?.Body == null)
+                {
+                    return;
+                }
 
-            var expectedStatementsCount = hasFinalizer ? 2 /*Dispose(true);GC.SuppressFinalize(this)*/: 1 /*Dispose(true)*/;
+                var parameterName = disposeMethod.ParameterList.Parameters.FirstOrDefault().Identifier.ToString();
+                if (CallsVirtualDispose(disposeMethod.Body, argument: parameterName))
+                {
+                    return;
+                }
 
-            if (disposeMethod.Body.ChildNodes().Count() != expectedStatementsCount ||
-                !CallsVirtualVoidDispose(disposeMethod, semanticModel, classSymbol) ||
-                (hasFinalizer && !CallsSuppressFinalize(disposeMethod, semanticModel)))
-            {
-                yield return new SecondaryLocation(disposeMethod.Identifier.GetLocation(),
-                    $"{classSymbol.Name}.Dispose() should contain only a call to {classSymbol.Name}.Dispose(true) and if the class contains a finalizer, call to GC.SuppressFinalize(this).");
+                AddLocation(disposeMethod.Identifier.GetLocation(),
+                    $"Modify Dispose({parameterName}) so that it calls base.Dispose({parameterName}).");
             }
 
-            var disposeMethodSymbol = semanticModel.GetDeclaredSymbol(disposeMethod);
-            if (disposeMethodSymbol == null)
+            private void VerifyDispose(MethodDeclarationSyntax disposeMethod, bool hasFinalizer)
             {
-                yield break;
-            }
+                if (disposeMethod?.Body == null)
+                {
+                    return;
+                }
 
-            if (disposeMethodSymbol.IsAbstract || disposeMethodSymbol.IsVirtual)
-            {
-                var modifier = disposeMethod.Modifiers.FirstOrDefault(t => t.IsKind(SyntaxKind.VirtualKeyword) || t.IsKind(SyntaxKind.AbstractKeyword));
-                yield return new SecondaryLocation(modifier.GetLocation(),
-                    $"{classSymbol.Name}.Dispose() should be sealed.");
-            }
+                var expectedStatementsCount = hasFinalizer
+                    ? 2  //// Dispose(true); GC.SuppressFinalize(this);
+                    : 1; //// Dispose(true);
 
-            if (disposeMethodSymbol.ExplicitInterfaceImplementations.Any())
-            {
-                yield return new SecondaryLocation(disposeMethod.Identifier.GetLocation(),
-                    $"{classSymbol.Name}.Dispose() should be public.");
-            }
-        }
+                if (disposeMethod.Body.ChildNodes().Count() != expectedStatementsCount ||
+                    !CallsVirtualDispose(disposeMethod.Body, argument: "true") ||
+                    (hasFinalizer && !CallsSuppressFinalize(disposeMethod.Body)))

```

Comment:
It feels like lines 181-187 could be simplified, but I could not come with anything better. Just extracting it to a method doesn't seem worthwhile.